### DISCUSSION TITLE
Ability to step frame by frame through footage in history

### DISCUSSION THREAD

**Discussion by drulia** [blakeblackshear/frigate#12448] [Ideas]
I'm new  to frigate, so maybe this was discussed, but I didn't find anything.

It would be great to have ability to step frame by frame in the review history video, sometimes it's only a frame or two that has captured details (plate number, face, etc.) and it's very hard to find it with only 1s increments. Have to export video and play on vlc to step through frames.

Also, it would be useful to have "Mark these items as reviewed" button at the top (at least on desktop), so that I wouldn't always need to scroll through all of videos to reach button. If I have people in the yard or some other action happening where I'm involved, I don't want to review what was captured and it's a bit slow to scroll on that page.

Using frigate 0.14 3'rd beta

**Comment by Mincka** [comment]
It was requested [there](https://github.com/blakeblackshear/frigate/discussions/11136#discussioncomment-9607623) but it will be more visible with your thread. I gave you a +1.

**Reply by hawkeye217** [reply]
Stepping frame by frame is not possible with HTML5 video in browsers. 

This is a limitation of the browser, not Frigate.

**Reply by Mincka** [reply]
That's sad since a function to do that has been deprecated.
Alternatively, would it possible to do sub-second seeking? It would not be frame by frame but more precise.

Finally, for frame by frame seeking, I tried this trick and it works fine in Edge and Chrome. It should also works in Safari, but not in Firefox.
https://github.com/angrycoding/requestVideoFrameCallback-prev-next?tab=readme-ov-file

**Reply by hawkeye217** [reply]
In order for us to consider implementing something like this, we would need a standard, well-supported solution that works in all browsers, not just a subset. 

Currently, the recommended method is to export and view it with VLC or another app as the OP has done.

**Reply by Mincka** [reply]
What about sub-second seeking?
5 to 10 fps are the recommended settings so it would be possible to use 0.2 or 0.1 second increments for up/down keys instead of 1 second. I tried in my JavaScript console and it works to get way more precise seeking.

**Reply by NickM-27** [reply]
Regardless of how it works by default it won't work for every use case. I wouldn't expect anything to change for 0.14.0, perhaps in the future there can be a "viewer mode" for when a user requires this level of granularity.

**Reply by Mincka** [reply]
Why it would not work? The stepping could take in account the fps setting of the camera to adjust it dynamically (0.05 second step for 20 fps, to 0.5s step for 2 fps for instance).
Even if you set it to 1 fps, it would just give you the current result.

**Reply by NickM-27** [reply]
> Why it would not work?

because if I want to skip 3 seconds forward I have to hit the button 30 times, this sub second granularity can not reasonably satisfy every use case someone might have for wanting to skip forward < 10 seconds.

**Reply by hawkeye217** [reply]
> Why it would not work? The stepping could take in account the fps setting of the camera to adjust it dynamically (0.05 second step for 20 fps, to 0.5s step for 2 fps for instance). Even if you set it to 1 fps, it would just give you the current result.

We have considered these things. Unfortunately this is not a well-supported solution in all browsers. 

https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/currentTime#usage_notes

**Reply by Mincka** [reply]
Do you see different results when you call `currentTime` for the same object? This does not seem to affect relative seeking.
Anyway, thank you for considering something around that. I still think that's relevant in the context of a NVR.
As I said previously, it would be very useful to be able to do that (in the browser) for events involving fast-passing cars for instance.

**Reply by hawkeye217** [reply]
We're not in disagreement that it would be a useful feature, we definitely think it would. The issue is that what you're proposing for a solution is not consistently supported across every browser.

As @NickM-27 said, we will not be making any changes to this for 0.14.0, but we are considering other options for upcoming versions.

**Reply by Mincka** [reply]
> > Why it would not work?
> 
> because if I want to skip 3 seconds forward I have to hit the button 30 times, this sub second granularity can not reasonably satisfy every use case someone might have for wanting to skip forward < 10 seconds.

You can just keep pushing the key to go faster, it already works like that.
Except that's actually less useful with 1 second steps than 0.x steps since you reach the equivalent of the 10 second jump in 2 seconds.

In my original post, I suggested to add buttons in the toolbar to do precision seeking so could have the "frame" seeking with the buttons and don't change the 1 second stepping with up / down. You would have the benefit of both.

**Reply by cbw** [reply]
> We have considered these things. Unfortunately this is not a well-supported solution in all browsers.
> 
> https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/currentTime#usage_notes

If I'm interpreting this correctly, seeking +0.1/-0.1s would work across browsers?

> For example, with reduced time precision, the result of video.currentTime will always be a multiple of 0.002, or a multiple of 0.1 (or privacy.resistFingerprinting.reduceTimerPrecision.microseconds) with privacy.resistFingerprinting enabled.

One potential and quite trivial implementation is to use the ctrl/meta modifier key alongside up/down. In `VideoControls.tsx` in the `onKeyboardShortcut` callback:

```
        case "ArrowDown":
          if (modifiers.ctrl) {
            onSeek(-0.1);
          } else {
            onSeek(-1);
          }
          break;
        case "ArrowUp":
          if (modifiers.ctrl) {
            onSeek(0.1);
          } else {
            onSeek(1);
          }
          break;
```

Created [a PR with this suggested implementation](https://github.com/blakeblackshear/frigate/pull/12656) for review.

**Reply by Mincka** [reply]
Very nice suggestion to use a modifier to make everyone happy.
May I ask why you put `1 / 30` in your PR? The code is not the same as above. Is it a mistake?
https://github.com/blakeblackshear/frigate/blob/054a576b7720081f8c80b5ca09219ed2fe241dac/web/src/components/player/VideoControls.tsx#L162

**Reply by ofcedwards** [reply]
> Created [a PR with this suggested implementation](https://github.com/blakeblackshear/frigate/pull/12656) for review.

I'm not sure how much more complex it would be to implement, but I prefer video players that have large skips (5+ seconds) while the video is playing and frame-by-frame while the video is paused. I'm assuming the intended use for your PR is with the video paused anyway. This would also allow for the feature to be used on mobile, where modifier keys aren't present. Overall, excellent idea.

**Reply by hawkeye217** [reply]
You're correct, mobile would not be supported with the code in that PR. As mentioned in the thread above, frame-by-frame is not possible in browsers.

We're looking at a more robust solution for precision and variable seeking that includes mobile support in a future version of Frigate.

**Comment by hawkeye217** [comment]
There have been many performance improvements since beta 3, including scrolling performance on the review page and the ability to select all review items with a keyboard shortcut. I’d recommend updating to beta 4.
